flink-connector-aws/flink-connector-dynamodb/src/test/java/org/apache/flink/connector/dynamodb/source/reader/PollingDynamoDbStreamsShardSplitReaderTest.java

```diff
@@ -0,0 +1,229 @@
+
+        RecordsWithSplitIds<Record> retrievedRecords = splitReader.fetch();
+
+        assertThat(retrievedRecords.nextRecordFromSplit()).isNull();
+        assertThat(retrievedRecords.nextSplit()).isNull();
+        assertThat(retrievedRecords.finishedSplits()).isEmpty();
+    }
+
+    @Test
+    void testAssignedSplitHasNoRecordsHandledGracefully() throws Exception {
+        TestDynamoDbStreamsProxy testStreamProxy = getTestStreamProxy();
+        PollingDynamoDbStreamsShardSplitReader splitReader =
+                new PollingDynamoDbStreamsShardSplitReader(testStreamProxy);
+
+        // Given assigned split with no records
+        String shardId = generateShardId(1);
+        testStreamProxy.addShards(shardId);
+        splitReader.handleSplitsChanges(
+                new SplitsAddition<>(Collections.singletonList(getTestSplit(shardId))));
+
+        // When fetching records
+        RecordsWithSplitIds<Record> retrievedRecords = splitReader.fetch();
+
+        // Then retrieve no records
+        assertThat(retrievedRecords.nextRecordFromSplit()).isNull();
+        assertThat(retrievedRecords.nextSplit()).isNull();
+        assertThat(retrievedRecords.finishedSplits()).isEmpty();
+    }
+
+    @Test
+    void testSingleAssignedSplitAllConsumed() throws Exception {
+        TestDynamoDbStreamsProxy testStreamProxy = getTestStreamProxy();
+        PollingDynamoDbStreamsShardSplitReader splitReader =
+                new PollingDynamoDbStreamsShardSplitReader(testStreamProxy);
+
+        // Given assigned split with records
+        String shardId = generateShardId(1);
+        testStreamProxy.addShards(shardId);
+        List<Record> expectedRecords =
+                Stream.of(getTestRecord("data-1"), getTestRecord("data-2"), getTestRecord("data-3"))
+                        .collect(Collectors.toList());
+        testStreamProxy.addRecords(
+                TestUtil.STREAM_ARN, shardId, Collections.singletonList(expectedRecords.get(0)));
+        testStreamProxy.addRecords(
+                TestUtil.STREAM_ARN, shardId, Collections.singletonList(expectedRecords.get(1)));
+        testStreamProxy.addRecords(
+                TestUtil.STREAM_ARN, shardId, Collections.singletonList(expectedRecords.get(2)));
+        splitReader.handleSplitsChanges(
+                new SplitsAddition<>(Collections.singletonList(getTestSplit(shardId))));
+
+        // When fetching records
+        List<Record> records = new ArrayList<>();
+        for (int i = 0; i < expectedRecords.size(); i++) {
+            RecordsWithSplitIds<Record> retrievedRecords = splitReader.fetch();
+            records.addAll(readAllRecords(retrievedRecords));
+        }
+
+        assertThat(records).containsExactlyInAnyOrderElementsOf(expectedRecords);
+    }
+
+    @Test
+    void testMultipleAssignedSplitsAllConsumed() throws Exception {
+        TestDynamoDbStreamsProxy testStreamProxy = getTestStreamProxy();
```

Comment:
I think this is a miss (for KDS too) where we should be adding multiple splits

Reply:
updated this. Added multiple splits